plugins/gitlab/shared/GitLabUtils.ts

```diff
@@ -0,0 +1,82 @@
+import queryString from "query-string";
+import env from "@shared/env";
+import { integrationSettingsPath } from "@shared/utils/routeHelpers";
+
+export class GitLabUtils {
+  public static clientId = env.GITLAB_CLIENT_ID;
+  public static oauthUrl = "https://gitlab.com/oauth";
+  public static apiBaseUrl = "https://gitlab.com/api/v4";
+
+  static get url() {
+    return integrationSettingsPath("gitlab");
+  }
+
+  /**
+   * @param error
+   * @returns URL to be redirected to upon authorization error from GitLab
+   */
+  public static errorUrl(error: string) {
+    return `${this.url}?error=${error}`;
+  }
+
+  /**
+   * @returns Callback URL configured for GitLab, to which users will be redirected upon authorization
+   */
+  public static callbackUrl(
+    { baseUrl, params }: { baseUrl: string; params?: string } = {
+      baseUrl: env.URL,
+      params: undefined,
+    }
+  ) {
+    return params
+      ? `${baseUrl}/api/gitlab.callback?${params}`
+      : `${baseUrl}/api/gitlab.callback`;
+  }
+
+  /**
+   * Generates the authorization URL for GitLab OAuth.
+   *
+   * @param {string} state - A unique state string to prevent CSRF attacks and maintain state between the request and callback.
+   * @returns {string} - The full URL to redirect the user to GitLab's OAuth authorization page.
+   */
+  static authUrl(state: string): string {
+    const baseUrl = `https://gitlab.com/oauth/authorize`;
+    const params = {
+      client_id: this.clientId,
+      redirect_uri: this.callbackUrl(),
+      response_type: "code",
+      state,
+      scope: "api read_api read_user",
```

Comment:
Is this the minimum scopes? I think `api` means write access 🤔

Reply:
Ah this scope is why it was failing for me as I didn't whitelist in application registration

Reply:
You're right. I kept hitting 401s so I added that, but I think it's cause I gave too many permissions to my application.

I'll take it out.